contract/scripts/snapshotPoSPool.js

```diff
@@ -1,36 +1,124 @@
+  nodes = nodes.filter(node => node.votes > 0);  // filter zero votes node
+  await getPoolStakersInfo(nodes);
+
+  fs.writeFileSync(path.join(__dirname, './PoSstakerSnapshot.json'), JSON.stringify(nodes, null, 2));
+  console.log('Finished');
+}
+
+main().catch(console.log);
+
+async function getPoSNodesFromScan() {
+  console.log("Fetching PoS node list from scan");
+  const LIMIT = 50;
+  let nodeList = [];
+  let skip = 0;
+  while(true) {
+    let url = `https://confluxscan.io/stat/list-pos-account?limit=${LIMIT}&orderBy=createdAt&reverse=true&skip=${skip}`;
+    let { status, body } = await superagent.get(url);
+    if (status !== 200) {
+      throw new Error(`Network error skip=${skip} limit=${LIMIT}`);
+    }
+    if (body.code !== 0) {
+      throw new Error(`API error skip=${skip} limit=${LIMIT}`);
+    }
+    nodeList = nodeList.concat(body.list.map(item => ({
+      posAddress: item.hex,
+    })));
+    if (body.list.length < LIMIT) break;
+    skip += 50;
   }
+  
+  return nodeList;
+}
 
-  // TODO: save data into json or csv
+async function getNodesBasicInfo(nodes) {
+  console.log('Fetching PoS nodes basic info');
+  for(let i in nodes) {
+    let node = nodes[i];
+    let powAddress = await PoSRegister.identifierToAddress(node.posAddress);
+    node.powAddress = powAddress;
+    let votes = await PoSRegister.getVotes(node.posAddress);
+    node.votes = votes[0] - votes[1]; // all PoS votes of this node, including locking, locked, unlocking
+  }
+}
 
-  console.log('Finished, total user: ', stakerNumber);
+async function getPoolStakersInfo(nodes) {
+  for(let i in nodes) {
+    let node = nodes[i];
+    let poolAddress = node.powAddress;
+    let addrInfo = address.decodeCfxAddress(poolAddress);
+    // check powAddress is contract
+    if (addrInfo.type.toLowerCase() !== 'contract') {
+      node.stakers = [];
+      continue;
+    }
+    // check powAddress has code
+    let code = await conflux.getCode(poolAddress);
+    if (code === '0x') {
+      node.stakers = [];
+      continue;
+    }
+    let stakers = await _getPoolStakers(poolAddress);
+    node.stakers = stakers;
+  }
 }
 
-main().catch(console.log);
\ No newline at end of file
+async function _getPoolStakers(poolAddress) {
+  console.log('Fetching stakers info for pool:', poolAddress);
+  try {
+    const pool = conflux.Contract({
+      abi: poolInfo.abi,
+      address: poolAddress,
+    });
+    let stakerNumber = await pool.stakerNumber();
+    let stakerInfo = [];
+    for(let i = 0; i < stakerNumber; i++) {
+      let addr = await pool.stakerAddress(i);
+      let userInfo = await pool.userSummary(addr);
+      let unlockingAndUnlocked = userInfo.votes - userInfo.available;
+      let unlocking = unlockingAndUnlocked - userInfo.unlocked;
+      stakerInfo.push({
+        address: addr,
+        votes: userInfo.available + unlocking,
+        available: userInfo.available,
+      });
+    }
+    return stakerInfo;
+  } catch(e) {
```

Comment:
What if any IO error occurred?

Strongly recommend to add retry mechanism, and fail the process if continous IO error happened. Otherwise, some stakers will be ignored for snapshot.

Reply:
If an error occurred, it will be logged out, user can rerun this script if the error is a network error